src/algos/fl_grid.py

```diff
@@ -0,0 +1,280 @@
+from collections import OrderedDict
+from typing import Any, Dict, List
+import torch
+import torch.nn as nn
+import random
+import numpy as np
+from math import ceil
+
+from algos.base_class import BaseFedAvgClient, BaseFedAvgServer
+
+from collections import defaultdict 
+from utils.stats_utils import from_round_stats_per_round_per_client_to_dict_arrays
+
+class FedGridClient(BaseFedAvgClient):
+    def __init__(self, config) -> None:
+        super().__init__(config)
+        
+    def get_collaborator_weights(self, reprs_dict, round):
+        """
+        Returns the weights of the collaborators for the current round
+        """ 
+        total_rounds = self.config["rounds"]
+        within_community_sampling = self.config.get("within_community_sampling",1)
+        p_within_decay = self.config.get("p_within_decay",None)
+        if p_within_decay is not None:
+            if p_within_decay == "linear_inc":         
+                within_community_sampling = within_community_sampling * (round / total_rounds)
+            elif p_within_decay == "linear_dec":
+                within_community_sampling = within_community_sampling * (1 - round / total_rounds)
+            elif p_within_decay == "exp_inc":
+                # Alpha scaled so that it goes from p to (1-p) in R rounds
+                alpha = np.log((1-within_community_sampling)/within_community_sampling)
+                within_community_sampling = within_community_sampling * np.exp(alpha * round / total_rounds)
+            elif p_within_decay == "exp_dec":
+                # Alpha scaled so that it goes from p to (1-p) in R rounds
+                alpha = np.log(within_community_sampling/(1-within_community_sampling))
+                within_community_sampling = within_community_sampling * np.exp(- alpha * round / total_rounds)
+            elif p_within_decay == "log_inc":
+                alpha = np.exp(1/within_community_sampling)-1
+                within_community_sampling = within_community_sampling * np.log2(1 + alpha * round / total_rounds)
+
+        self.grid_size = int(self.config["num_clients"]**0.5)
+
+        self.num_clients = self.config["num_clients"]
+
+        selected_ids = []
+
+        # Left
+        if self.node_id % self.grid_size != 1:
+            selected_ids.append(self.node_id - 1)
+        
+        # Right
+        if self.node_id % self.grid_size != 0 and self.node_id < self.num_clients:
```

Comment:
In which situation will the second part of this `if` statement be false?

Reply:
nevermind, figured. Make it `self.node_id == self.num_clients` to be more explicit about it